keps/sig-network/2091-admin-network-policy/README.md

```diff
@@ -0,0 +1,1116 @@
+  // operation. If this field is empty or missing, this rule matches no sources. 
+  // If this field is present and contains at least one item, this rule
+  // allows/denies/passes traffic from the defined AdminNetworkPolicyPeer
+  From      []AdminNetworkPolicyPeer
+}
+
+
+type AdminNetworkPolicyEgressRule struct {
+  // Name is an identifier for this rule. Must be unique within an instance of ANP. Can be
+  // same across policy instances.
+  // +optional
+  Name     string
+  // Action specifies whether this rule must pass, allow or deny traffic. 
+  // Allow: explicitly allows the selected traffic 
+  // Deny: explicitly denies the selected traffic
+  // Pass: explicitly allows the selected traffic to skip and remaining AMP rules 
+  // and be delegated by K8's Network Policy. 
+  Action   RuleAction
+  // List of ports for outgoing traffic.
+  // Each item in this list is combined using a logical OR. If this field is
+  // empty or missing, this rule matches no ports (traffic not restricted by port).
+  // If this field is present and contains at least one item, then this rule 
+  // allows/denies/passes traffic only if the traffic matches at least one port in the list.
+  // +optional
+  Ports    []networkingv1.NetworkPolicyPort
+  // List of destinations to which traffic will be allowed/denied/passed from the entities 
+  // selected by this AdminNetworkPolicyRule. Items in this list are combined using a logical OR 
+  // operation. If this field is empty or missing, this rule matches no destinations. 
+  // If this field is present and contains at least one item, this rule
+  // allows/denies/Passes traffic to the defined AdminNetworkPolicyPeer
+  // +optional
+  To       []AdminNetworkPolicyPeer
+}
+
+const (
+  // RuleActionPass enables admins to provide exceptions to ClusterNetworkPolicies and delegate this rule to
+  // K8s NetworkPolicies.
+  RuleActionPass    RuleAction = "Pass"
+  // RuleActionDeny enables admins to deny specific traffic.
+  RuleActionDeny    RuleAction = "Deny"
+  // RuleActionAllow enables admins to specifically allow certain traffic.
+  RuleActionAllow   RuleAction = "Allow"
+)
+
+// Exactly one of the selector pointers should be set in a given AdminNetworkPolicyPeer
+type AdminNetworkPolicyPeer struct {
+  NamespaceSetSelector         *NamespaceSet
+  NamespaceSetAndPodSelector   *NamespaceSetAndPod
+}
+
+// NamespaceSet defines a flexible way to select Namespaces in a cluster.
+type NamespaceSet struct {
+  Scope               NamespaceMatchType
+  // Labels should be set only when scope is "SameLabels" or "NotSameLabels.
+  // +optional
+  Labels              []string
+  // NamespaceSelector should be set only when scope is "Selector". 
+  NamespaceSelector   *metav1.LabelSelector
+}
+
+// NamespaceSetAndPod defines a way to select Namespaces and Pods in a cluster 
+type NamespaceSetAndPod struct {
+  Scope               NamespaceMatchType
```

Comment:
I want to play with this a little bit because I find it makes it a little bit hard to wrap my head around some of the policy rules. Maybe that's just me, though.

Maybe some of my mental block would be alleviated if this field were "NamespaceScope" or "NamespaceMatchType". Scope just feels too vague maybe?

Reply:
The alternative might be to get rid of this field altogether and encode it into the namespace selector itself:



```yaml
// Match all namespaces which share my label value for the label "tenant"
namespaceSelector:
  sharesLabels:
    - tenant
```

```yaml
// Match all namespaces which do not share my label value for the label "tenant"
namespaceSelector:
  notSharesLabels:
    - tenant
```

```yaml
// Match the local pod's namespace (i.e., by default pod selectors are scope to their own namespace)
namespaceSelector: nil
```